internal/integrationtest/util.go

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integrationtest
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/raw"
+	peerbind "go.uber.org/yarpc/peer"
+	"go.uber.org/yarpc/peer/roundrobin"
+)
+
+// TransportSpec specifies how to create test clients and servers for a transport.
+type TransportSpec struct {
+	NewServerTransport func(t *testing.T, addr string) peer.Transport
+	NewClientTransport func(t *testing.T) peer.Transport
+	NewInbound         func(x peer.Transport, addr string) transport.Inbound
+	NewUnaryOutbound   func(x peer.Transport, pc peer.Chooser) transport.UnaryOutbound
+	Identify           func(addr string) peer.Identifier
+	Addr               func(x peer.Transport, ib transport.Inbound) string
+}
+
+// Test runs reusable tests with the transport spec.
+func (s TransportSpec) Test(t *testing.T) {
+	t.Run("reuse connection with round robin", s.TestReuseConnectionWithRoundRobin)
+	t.Run("backoff reconnection with round robin", s.TestBackoffWithRoundRobin)
+	t.Run("reconnect using round robin", s.TestReconnectWithRoundRobin)
+	t.Run("lose patience connecting with round robin", s.TestLosePatienceWithRoundRobin)
+}
+
+// NewClient returns a running dispatcher and a raw client for the echo
+// procedure.
+func (s TransportSpec) NewClient(t *testing.T, addrs []string) (*yarpc.Dispatcher, raw.Client) {
+	// Convert peer addresses into peer identifiers for a peer list.
+	ids := make([]peer.Identifier, len(addrs))
+	for i, addr := range addrs {
+		ids[i] = s.Identify(addr)
+	}
+
+	x := s.NewClientTransport(t)
+
+	pl := roundrobin.New(x)
```

Comment:
all the variable names - I know golang leans towards short variable names, but when I read the code, I natively do not know what `pl, pc, ob, d` are